File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/function/system/array/IndexOfArray.java
package com.fincity.nocode.kirun.engine.function.system.array;

import java.util.List;
import java.util.Map;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.model.EventResult;
import com.fincity.nocode.kirun.engine.model.FunctionOutput;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.fincity.nocode.kirun.engine.util.primitive.PrimitiveUtil;
import com.google.gson.JsonArray;
import com.google.gson.JsonPrimitive;

import reactor.core.publisher.Mono;

public class IndexOfArray extends AbstractArrayFunction {

	public IndexOfArray() {
		super("IndexOfArray", List.of(PARAMETER_ARRAY_SOURCE, PARAMETER_ARRAY_SECOND_SOURCE, PARAMETER_INT_FIND_FROM),
		        EVENT_RESULT_INTEGER);
	}

	@Override
	protected Mono<FunctionOutput> internalExecute(ReactiveFunctionExecutionParameters context) {

		JsonArray source = context.getArguments()
		        .get(PARAMETER_ARRAY_SOURCE.getParameterName())
		        .getAsJsonArray();

		JsonArray secondSource = context.getArguments()
		        .get(PARAMETER_ARRAY_SECOND_SOURCE.getParameterName())
		        .getAsJsonArray();

		JsonPrimitive from = context.getArguments()
		        .get(PARAMETER_INT_FIND_FROM.getParameterName())
		        .getAsJsonPrimitive();

		if (source.isEmpty() || secondSource.isEmpty())
			return Mono.just(new FunctionOutput(
			        List.of(EventResult.outputOf(Map.of(EVENT_RESULT_INTEGER.getName(), new JsonPrimitive(-1))))));

		if (from.getAsInt() < 0 || from.getAsInt() > source.size() || source.size() < secondSource.size())
			throw new KIRuntimeException("Given from second source is more than the size of the source array");

		int len = from.getAsInt();
		int secondSourceSize = secondSource.size();
		int index = -1;

		for (int i = len; i < source.size(); i++) {
			int j = 0;

			if (PrimitiveUtil.compare(source.get(i), secondSource.get(j)) == 0) {
				while (j < secondSourceSize) {
					if (PrimitiveUtil.compare(source.get(i + j), secondSource.get(j)) != 0) {
						break;
					}
					j++;
				}
				if (j == secondSourceSize) {
					index = i;
					break;
				}
			}
		}

		return Mono.just(new FunctionOutput(
		        List.of(EventResult.outputOf(Map.of(EVENT_RESULT_INTEGER.getName(), new JsonPrimitive(index))))));
	}
}

File: kirun-js/src/engine/function/system/array/IndexOfArray.ts
import { KIRuntimeException } from '../../../exception/KIRuntimeException';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { PrimitiveUtil } from '../../../util/primitive/PrimitiveUtil';
import { AbstractArrayFunction } from './AbstractArrayFunction';

export class IndexOfArray extends AbstractArrayFunction {
    public constructor() {
        super(
            'IndexOfArray',
            [
                IndexOfArray.PARAMETER_ARRAY_SOURCE,
                IndexOfArray.PARAMETER_ARRAY_SECOND_SOURCE,
                IndexOfArray.PARAMETER_INT_FIND_FROM,
            ],
            IndexOfArray.EVENT_RESULT_INTEGER,
        );
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let source: any[] = context
            ?.getArguments()
            ?.get(IndexOfArray.PARAMETER_ARRAY_SOURCE.getParameterName());

        let secondSource: any[] = context
            ?.getArguments()
            ?.get(IndexOfArray.PARAMETER_ARRAY_SECOND_SOURCE.getParameterName());

        let from: number = context
            ?.getArguments()
            ?.get(IndexOfArray.PARAMETER_INT_FIND_FROM.getParameterName());

        if (source.length == 0 || secondSource.length == 0)
            return new FunctionOutput([
                EventResult.outputOf(new Map([[IndexOfArray.EVENT_RESULT_INTEGER.getName(), -1]])),
            ]);

        if (from < 0 || from > source.length || source.length < secondSource.length)
            throw new KIRuntimeException(
                'Given from second source is more than the size of the source array',
            );

        let secondSourceSize: number = secondSource.length;
        let index: number = -1;

        for (let i: number = from; i < source.length; i++) {
            let j: number = 0;
            if (PrimitiveUtil.compare(source[i], secondSource[j]) == 0) {
                while (j < secondSourceSize) {
                    if (PrimitiveUtil.compare(source[i + j], secondSource[j]) != 0) {
                        break;
                    }
                    j++;
                }
                if (j == secondSourceSize) {
                    index = i;
                    break;
                }
            }
        }

        return new FunctionOutput([
            EventResult.outputOf(new Map([[IndexOfArray.EVENT_RESULT_INTEGER.getName(), index]])),
        ]);
    }
}
